File: adapters/android/starbidz-levelplay/src/main/kotlin/io/starbidz/levelplay/StarbidBannerAdapter.kt
package io.starbidz.levelplay

import android.app.Activity
import android.view.View
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.FrameLayout
import com.ironsource.mediationsdk.ISBannerSize
import com.ironsource.mediationsdk.adunit.adapter.BaseBanner
import com.ironsource.mediationsdk.adunit.adapter.listener.BannerAdListener
import com.ironsource.mediationsdk.adunit.adapter.utility.AdData
import io.starbidz.core.AdFormat
import io.starbidz.core.AdResult
import io.starbidz.core.StarbidzAd
import io.starbidz.core.Starbidz
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class StarbidBannerAdapter : BaseBanner<BannerAdListener>() {

    private var webView: WebView? = null
    private var container: FrameLayout? = null
    private var ad: StarbidzAd? = null

    override fun loadAd(
        adData: AdData,
        activity: Activity,
        bannerSize: ISBannerSize,
        listener: BannerAdListener
    ) {
        val placementId = adData.getString(StarbidCustomAdapter.PARAM_PLACEMENT_ID)

        if (placementId.isNullOrEmpty()) {
            listener.onAdLoadFailed("INVALID_CONFIGURATION", "Missing placementId")
            return
        }

        val width = bannerSize.width
        val height = bannerSize.height

        CoroutineScope(Dispatchers.Main).launch {
            when (val result = Starbidz.requestBid(activity, placementId, AdFormat.BANNER, width, height)) {
                is AdResult.Success -> {
                    ad = result.ad
                    createBannerView(activity, result.ad, width, height, listener)
                }
                is AdResult.Error -> {
                    listener.onAdLoadFailed("NO_FILL", result.message)
                }
                is AdResult.NoBid -> {
                    listener.onAdLoadFailed("NO_FILL", "No bid")
                }
            }
        }
    }

    private fun createBannerView(
        activity: Activity,
        ad: StarbidzAd,
        width: Int,
        height: Int,
        listener: BannerAdListener
    ) {
        container = FrameLayout(activity).apply {
            layoutParams = FrameLayout.LayoutParams(
                dpToPx(activity, width),
                dpToPx(activity, height)
            )
        }

        webView = WebView(activity).apply {
            layoutParams = FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT
            )
            settings.javaScriptEnabled = true
            settings.domStorageEnabled = true

            webViewClient = object : WebViewClient() {
                override fun onPageFinished(view: WebView?, url: String?) {
                    super.onPageFinished(view, url)
                    listener.onAdLoaded(container!!)
                    listener.onAdShown()

                    CoroutineScope(Dispatchers.IO).launch {
                        Starbidz.trackImpression(ad.bidId, "")
                    }
                }
            }
        }

        container?.addView(webView)

        when (ad.creative.type) {
            StarbidzAd.CreativeType.HTML -> {
                webView?.loadDataWithBaseURL(null, wrapHtml(ad.creative.content), "text/html", "UTF-8", null)
            }
            StarbidzAd.CreativeType.IMAGE -> {
                val imgHtml = """
                    <html><body style="margin:0;padding:0;">
                    <img src="${ad.creative.content}" style="width:100%;height:100%;object-fit:contain;"/>
                    </body></html>
                """.trimIndent()
                webView?.loadDataWithBaseURL(null, imgHtml, "text/html", "UTF-8", null)
            }
            else -> {
                listener.onAdLoadFailed("INTERNAL_ERROR", "Unsupported creative type")
            }
        }
    }

    override fun destroyAd(adData: AdData) {
        webView?.destroy()
        webView = null
        container?.removeAllViews()
        container = null
        ad = null
    }

    private fun wrapHtml(content: String) = """
        <!DOCTYPE html><html><head>
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <style>* { margin: 0; padding: 0; }</style>
        </head><body>$content</body></html>
    """.trimIndent()

    private fun dpToPx(activity: Activity, dp: Int): Int {
        return (dp * activity.resources.displayMetrics.density).toInt()
    }
}

File: adapters/ios/StarbidLevelPlay/Sources/StarbidBannerAdapter.swift
import Foundation
import UIKit
import WebKit
import StarbidCore

public class StarbidBannerAdapter: NSObject {

    public weak var delegate: ISBannerAdDelegate?

    private var webView: WKWebView?
    private var containerView: UIView?
    private var ad: StarbidAd?
    private var config: ISAdapterConfig?

    public override init() {
        super.init()
    }

    public func loadAd(config: ISAdapterConfig, size: CGSize, delegate: ISBannerAdDelegate) {
        self.config = config
        self.delegate = delegate

        StarbidCustomAdapter.initializeSDK(config: config)

        guard let placementId = config.getString(StarbidCustomAdapter.paramPlacementId), !placementId.isEmpty else {
            delegate.adDidFailToLoadWithError("Missing placementId in configuration")
            return
        }

        let width = Int(size.width)
        let height = Int(size.height)

        Starbidz.shared.requestBid(
            placementId: placementId,
            format: .banner,
            width: width,
            height: height
        ) { [weak self] result in
            guard let self = self else { return }

            DispatchQueue.main.async {
                switch result {
                case .success(let ad):
                    self.ad = ad
                    self.createBannerView(ad: ad, size: size)

                case .error(let message, _):
                    self.delegate?.adDidFailToLoadWithError(message)

                case .noBid:
                    self.delegate?.adDidFailToLoadWithError("No fill")
                }
            }
        }
    }

    private func createBannerView(ad: StarbidAd, size: CGSize) {
        containerView = UIView(frame: CGRect(origin: .zero, size: size))
        containerView?.backgroundColor = .clear

        let config = WKWebViewConfiguration()
        config.allowsInlineMediaPlayback = true

        webView = WKWebView(frame: containerView!.bounds, configuration: config)
        webView?.navigationDelegate = self
        webView?.scrollView.isScrollEnabled = false
        webView?.isOpaque = false
        webView?.backgroundColor = .clear
        webView?.autoresizingMask = [.flexibleWidth, .flexibleHeight]

        containerView?.addSubview(webView!)

        switch ad.creative.type {
        case .html:
            let html = wrapHtml(ad.creative.content)
            webView?.loadHTMLString(html, baseURL: nil)

        case .image:
            let imgHtml = """
            <html>
            <body style="margin:0;padding:0;">
            <img src="\(ad.creative.content)" style="width:100%;height:100%;object-fit:contain;"/>
            </body>
            </html>
            """
            webView?.loadHTMLString(imgHtml, baseURL: nil)

        case .vast:
            delegate?.adDidFailToLoadWithError("VAST not supported for banner")
        }
    }

    public func destroyAd() {
        webView?.stopLoading()
        webView?.removeFromSuperview()
        webView = nil
        containerView?.removeFromSuperview()
        containerView = nil
        ad = nil
    }

    private func wrapHtml(_ content: String) -> String {
        return """
        <!DOCTYPE html>
        <html>
        <head>
            <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">
            <style>
                * { margin: 0; padding: 0; box-sizing: border-box; }
                html, body { width: 100%; height: 100%; overflow: hidden; }
            </style>
        </head>
        <body>
            \(content)
        </body>
        </html>
        """
    }
}

// MARK: - WKNavigationDelegate

extension StarbidBannerAdapter: WKNavigationDelegate {

    public func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        guard let container = containerView else { return }
        delegate?.adDidLoad(container)
        delegate?.adDidShow()

        if let ad = ad {
            Starbidz.shared.trackImpression(bidId: ad.bidId, placementId: "")
        }
    }

    public func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        delegate?.adDidFailToLoadWithError(error.localizedDescription)
    }

    public func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {

        if navigationAction.navigationType == .linkActivated {
            delegate?.adDidClick()

            if let ad = ad {
                Starbidz.shared.trackClick(bidId: ad.bidId, placementId: "")
            }

            if let url = navigationAction.request.url {
                UIApplication.shared.open(url)
            }

            decisionHandler(.cancel)
            return
        }

        decisionHandler(.allow)
    }
}
